convert2rhel/backup.py

```diff
@@ -104,6 +108,150 @@ def restore_pkgs(self):
+        """
+        restorables = self._restorables
+
+        if not restorables:
+            raise IndexError("pop from empty list")
+
+        # We want to restore in the reverse order the changes were enabled.
+        for restorable in reversed(restorables):
+            restorable.restore()
+
+        # Reset the internal storage in case we want to use it again
+        self._restorables = []
+
+        # Now that we know everything succeeded, reverse the list that we return to the user
+        restorables.reverse()
+
+        return restorables
+
+
+@six.add_metaclass(abc.ABCMeta)
+class RestorableChange(object):
+    """
+    Interface definition for types which can be restored.
+    """
+
+    @abc.abstractmethod
+    def __init__(self):
+        self.enabled = False
+
+    @abc.abstractmethod
+    def enable(self):
+        """
+        Backup should be idempotent.  In other words, it should know if the resource has already
+        been backed up and refuse to do so a second time.
+        """
+        self.enabled = True
+
+    @abc.abstractmethod
+    def restore(self):
+        """
+        Restore the state of the system.
+        """
+        self.enabled = False
+
+
+class RestorableRpmKey(RestorableChange):
+    def __init__(self, keyfile):
+        super(RestorableRpmKey, self).__init__()
+        self.previous_state = None
+        self.keyfile = keyfile
+        self.keyid = utils.find_keyid(keyfile)
+
+    def enable(self):
+
+        # For idempotence, do not back this up if we've already done so.
+        if self.enabled:
+            return
+
+        if not self.installed:
+            output, ret_code = utils.run_subprocess(["rpm", "--import", self.keyfile], print_output=False)
+            if ret_code != 0:
+                raise utils.ImportGPGKeyError("Failed to import the GPG key %s: %s" % (self.keyfile, output))
+
+            self.previous_state = "not_installed"
+            loggerinst.info("GPG key %s imported", self.keyid)
+
+        else:
+            self.previous_state = "installed"
+
+        super(RestorableRpmKey, self).enable()
+
+    @property
+    def installed(self):
+        output, status = utils.run_subprocess(["rpm", "-q", "gpg-pubkey-%s" % self.keyid], print_output=False)
+
+        if status == 0:
+            return True
+
+        if status == 1 and "package gpg-pubkey-%s is not installed" % self.keyid in output:
+            return False
+
+        raise utils.ImportGPGKeyError(
+            "Searching the rpmdb for the gpg key %s failed: Code %s: %s" % (self.keyid, status, output)
+        )
+
+    def restore(self):
+        if not self.enabled or self.previous_state in (None, "installed"):
+            self.enabled = False
+            return
+
+        # The self.previous_state == "not_installed" case
+        utils.run_subprocess(["rpm", "-e", "gpg-pubkey-%s" % self.keyid])
+
+        super(RestorableRpmKey, self).restore()
```

Comment:
@abadger I see that this whole class is missing the docstrings here. I can see why `restore` and `enabled` might not have one, it makes sense. But for the class itself and the `installed`, I would say that it should be good to at least have a small docstring to indicate what it is.

Reply:
Docstrings now added.